### DISCUSSION TITLE
Testing with cookies

### DISCUSSION THREAD

**Discussion by mattkibbler** [livewire/livewire#4129] [7. Bugs (archived)]
Hi all,

I've ran into an issue when testing a component that accesses cookies.

Using `Livewire::test('component-name')->withCookie('cookie-name', 'cookie-value')`, like you would with usual Laravel HTTP tests, doesn't throw an error so I assumed it would work but it doesn't seem to actually do anything.

It says in the docs that Livewire components are stateless so perhaps this makes sense, but could I get some clarification on whether it should be working or not?

Thanks!

**Comment by joshhanley** [comment]
@mattkibbler can you share an example of a test that uses that so we can have a look?

Thanks!

**Reply by mattkibbler** [reply]
Hi @joshhanley.

Sure, I've set up a basic example repo [here](https://github.com/mattkibbler/livewire-cookie-test).

The basic idea is that I have a shopping basket model with this method to return an existing basket from an ID reference in a cookie, or create a new one:

```php
public static function findOrCreate() 
{

    $basket = null;
    $basketId = Cookie::get('basket');

    // Try to get an existing basket based on the stored ID
    if($basketId) {
        $basket = Basket::find($basketId);
    }

    // If no basket found, make a new one
    if(!$basket) {
        $basket = Basket::create();
    }

    $oneYearInMinutes = 525600;
    Cookie::queue('basket', $basket->id, $oneYearInMinutes);

    return $basket;

}
```

I can then reference the current user's basket wherever, for example in a Livewire component:

```php
public function render()
{

    $basket = Basket::findOrCreate();

    return view('livewire.cookie-component', [
        'basketId' => $basket->id,
    ]);
}

public function submit() 
{

    $basket = Basket::findOrCreate();

    // Do basket/checkout stuff

}
```

This works fine in the browser, but I can't test the Livewire components reliably because the test component instance don't seem to be aware of cookies.

I've set up two tests in the example repo, both using the Laravel request `withCookie` method.

The first test asserts that the basket ID is always the same and this fails.

```php
public function test_cookies_work_in_livewire_component_tests()
{

    // Basket ID will be 1
    $basket = Basket::factory()->create();

    Livewire::test('cookie-component')
        ->withCookie('basket', $basket->id)
        ->assertSee('Basket #1')
        ->call('$refresh')
        ->assertSee('Basket #1');

}
```

The second test passes because it asserts that the basket ID increments with each call to `Basket::findOrCreate`, demonstrating that the Livewire test doesn't know about the basket ID in the cookie.

```php
public function test_cookies_dont_work_in_livewire_component_tests()
{

    // Basket ID will be 1
    $basket = Basket::factory()->create();

    Livewire::test('cookie-component')
        ->withCookie('basket', $basket->id)
        ->assertSee('Basket #2')
        ->call('$refresh')
        ->assertSee('Basket #3');

}
```

**Comment by joshhanley** [comment]
@mattkibbler thanks! Yeah the reason I don't think that will work currently, is because Livewire doesn't have a `withCookie` method.

So it will use Laravel's `withCookie` method instead which possibly isn't compatible with Livewire.

The reason I don't think it will work, it because when testing Laravel you would be faking a network request by calling `$this->get()`, for example,  so you can add cookies to that network request. Where as with Livewire it doesn't fake the network request, it only fakes the lifecycle of a Livewire component.

This might be able to be changed, and we'd certainly welcome a PR to add that if you are up to it.

But before that, my question for you is, why are you using a cookie to keep track of their basket ID, instead of putting it in the session?

Hope this helps!

**Reply by mattkibbler** [reply]
Thanks, @joshhanley.

I'm using cookies because I want the shopping basket to persist for a long time so when a user returns after a week or a month their basket contents are still there, but I don't want to increase the session lifetime that much.

Not sure if I'm up to implementing this functionality myself, but I may give it a go when I have a moment. 

I see in the `TestableLivewire` class that any methods not supported are passed to the `lastResponse`, which is an instance of `Illuminate\Testing\TestResponse`:

```php
public function __call($method, $params)
{
    if (static::hasMacro($method)) {
        return $this->macroCall($method, $params);
    }

    $this->lastResponse->$method(...$params);
    
    return $this;
}
```

As far as I can see this is to allow for the `withQueryParams` method to be used. If this is the case, could this method be whitelisted and any others that may not be supported throw an exception?

**Comment by edalzell** [comment]
There will be when https://github.com/livewire/livewire/pull/6822 gets merged.
